Introduce first version of ECDSA adaptor signature spec (#114)
discreetlogcontracts/dlcspecs: ECDSA-adaptor.md

In our context this is an oracle signature *s* value.

Note that if this function returns successfully and `a` has been verified with `ecdsa_adaptor_verify` for some public key and message hash then this implies that `sig` is a valid ECDSA signature on the message hash
i.e. there is no strict need to verify `sig` after (or before) successfully calling `ecdsa_adaptor_recover` if you have already ensured that `ecdsa_adaptor_verify` passes on the original adaptor signature.
However, successfully recovering the signature does not ensure that it follows the `low_s` rule described in [[BIP62]] necessary to be able to broadcast the signature on a transaction in Bitcoin.
This point is not relevant to on-chain DLCs since the ECDSA signatures always come from the blockchain where they have already been verified.

## Specification tests

The test vectors for this specification are in [./test/ecdsa_adaptor.json] and come in two kinds:

- `verification`: These test the verifier in three stages:
  - Check `adaptor_sig` passes `ecdsa_adaptor_verify`.
  - Check that `ecdsa_adaptor_decrypt` yields `signature`.
  - Check that  `ecdsa_adaptor_recover` recovers `decryption_key`.

> - `recovery`: These test only the key recovery function. The recovery function should recovery the correct decryption key.

Reviewer: ```suggestion
- `recovery`: These test only the key recovery function. The recovery function should recover the correct decryption key.
```

# ECDSA adaptor signatures

DLCs work by conditioning a transaction on the revelation of an oracle's signature .
Concretely the approach is to encrypt a transaction signature such that the oracle's released signature is the decryption key.
Once the oracle has revealed a signature, either party may use it to decrypt the corresponding transaction signature and broadcast the transaction.
By this method an oracle's view of a real world event enforces a particular distribution of funds between two parties according to their private bet.

## Adaptor Signatures

*Adaptor signatures* are an efficient form of verifiable signature encryption that can achieve our required structure.

> Given a anticipated signature `Y` we may encrypt a signature (for the purposes of this document, an ECDSA signature) under `Y` such that once a party learns the discrete logarithm of `Y` they can decrypt it and get a valid ECDSA signature.

Reviewer: ```suggestion
Given an anticipated signature `Y` we may encrypt a signature (for the purposes of this document, an ECDSA signature) under `Y` such that once a party learns the discrete logarithm of `Y` they can decrypt it and get a valid ECDSA signature.
```

> Given a anticipated signature `Y` we may encrypt a signature (for the purposes of this document, an ECDSA signature) under `Y` such that once a party learns the discrete logarithm of `Y` they can decrypt it and get a valid ECDSA signature.

Reviewer: nit: `Y` isn't a signature per se, bur rather a commitment to a signature, or a public key derived from an anticipated signature

Author: I think `Y` should just be "public key"? Just a typo I think.

Author: The name I use for the commitment/public key is "anticipated signature". It came from  Tadge: https://diyhpl.us/wiki/transcripts/discreet-log-contracts/.

I *think* I am using it correctly here.

Reviewer: From that transcript, it looks like it's ambiguous as to whether "Anticipated Signature" refers to `Y` or just the scheme in general. However, my point might be a little pedantic and I don't think it matters all that much, and I can see the utility of having a shorthand like "Anticipated Signature."

> An interesting aspect of adaptor signatures is that they leak the decryption key (the discrete logarithm of `Y`) when they are decrypted.

Reviewer: It might help to be a little more explicit here:
```suggestion
An interesting aspect of adaptor signatures is that the decryption key (the discrete logarithm of the encryption key `Y`) can be derived if both the plaintext (the valid ECDSA signature) and the ciphertext (the adaptor signature) are known.
```

An appropriate ECDSA adaptor signature scheme that can be used in Bitcoin prior to the Taproot soft-fork was originally posted to the Bitcoin mailing list by [[Mor]] and then further refined by [[Fou],[Aum]].
The algorithms presented here resemble the scheme from [[Fou]] but the security proof of the scheme in [[Aum]] can be applied to it which requires a proof of knowledge of the discrete logarithm to be attached to each encryption key.
Although we do not explicitly attach proofs of knowledge, in the DLC application the encryption key is the anticipated signature of an oracle which by definition the oracle must "know" and therefore we can safely omit the proof of knowledge.

## Notation and Conventions

- secp256k1 operations: We refer to *points* and *scalars* of the secp256k1 group and the operations between them.
  - *n* denotes the curve order of secp256k1 `0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141`
  - Scalars are integers between *0..(n-1)* inclusive.
  - Scalars have three infix operations: multiplication (`*`), addition (`+`) and subtraction (`-`) which are done modulo *n*.

>   - Scalars have a postfix operation (`⁻¹`) which denotes inversion modulo *n*.

Reviewer: ```suggestion
  - Scalars have a postfix operation (`⁻¹`) which denotes multiplicative inversion modulo *n*.
```

  - Points are members of the secp256k1 group.
  - Points have two binary operations between them: addition (`+`) and subtraction (`-`) which represent the group operation and the inverse group operation respectively.
  - Points and scalars have a infix multiplication (`*`) operation which for `s * P` means adding the point `P` to itself `s` times.

>   - *G* is the standard generator point of secp256k1 with coordinates:

Reviewer: ```suggestion
  - *G* is the generator point of secp256k1 with coordinates:
```

- Return `proof`

### Verifying

The `DLEQ_verify((X,Y,Z), proof)` algorithm takes as input:

- `(X,Y,Z)`: Three **non-zero** secp256k1 points which define the statement we are verifying.
             Specifically, there exists `x` such that `X = x * G` and `Z = x * Y`.
- `proof`: the proof for the statement.

and is defined as:

- Parse `proof` as two scalars `(b,c)`
- Set `A_G` to `c * G - b * X`
- Set `A_Y` to `c * Y - b * Z`
- Set `implied_b` to `H(X || Y || Z || A_G || A_Y)`
- Check `implied_b == b`

## Adaptor Signature

### Encrypted Signing

The encrypted signing algorithm `ecdsa_adaptor_encrypt(x, Y, message_hash)` takes as input:

- `x`: The signing *secret key*.
- `Y`: The encryption *public key* (in our case the *anticipated signature* of an oracle associated with a particular outcome).
- `message_hash`: A 32-byte array which **must** be the hash of the message (in our case a transaction digest).

and is defined as:

- Set `k` to `sample_nonce(Y || message_hash || x)`
- Set `m` to `scalar(message_hash)`
- Set `R_a` to `k * G`
- Set `R` to `k * Y`
- Set `proof` to `DLEQ_prove(k, (R_a, Y, R))`
- Set `r` to the x-coordinate of `R` modulo `n` (i.e. convert it to a scalar)

> - Set `s_a` to `k^-1 (m + r * x)`

Reviewer: ```suggestion
- Set `s_a` to `k⁻¹ (m + r * x)`
```
